include/nanorange/algorithm/sample.hpp

```diff
@@ -0,0 +1,118 @@
+// nanorange/algorithm/sample.hpp
+//
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef NANORANGE_ALGORITHM_SAMPLE_HPP_INCLUDED
+#define NANORANGE_ALGORITHM_SAMPLE_HPP_INCLUDED
+
+#include <nanorange/ranges.hpp>
+#include <nanorange/random.hpp>
+#include <nanorange/algorithm/min.hpp>
+
+#include <random>
+
+NANO_BEGIN_NAMESPACE
+
+namespace detail {
+
+struct sample_fn {
+private:
+    template <typename I, typename S, typename O, typename Gen>
+    static O impl_fwd(I first, S last, O out, iter_difference_t<I> n, Gen&& g)
+    {
+        using diff_t = iter_difference_t<I>;
+        using distr_t = std::uniform_int_distribution<diff_t>;
+        using param_t = typename distr_t::param_type;
+
+        distr_t D;
+
+        auto unsampled_size = nano::distance(first, last);
+
+        for (n = nano::min(n, unsampled_size); n != 0; ++first) {
+            if (D(g, param_t(0, --unsampled_size)) < n) {
+                *out++ = *first;
+                --n;
+            }
+        }
+
+        return out;
+    }
+    template <typename I, typename S, typename O, typename Gen>
+    static O impl_ra(I first, S last, O out, iter_difference_t<I> n, Gen&& g) {
+        using diff_t = iter_difference_t<I>;
+        using distr_t = std::uniform_int_distribution<diff_t>;
+        using param_t = typename distr_t::param_type;
+
+        distr_t D;
+        diff_t k = 0;
+
+        for(; first != last && k < n; ++first, (void) ++k) {
+            out[k] = *first;
+        }
+
+        diff_t size = k;
+        for (; first != last; ++first, (void) ++k) {
+            diff_t r = distr_t(0, k)(g);
+            if (D(g, param_t(0, k)) < size) {
+                out[r] = *first;
+            }
+        }
+
+        return out + nano::min(n, k);
+    }
+
+    template <typename I, typename S, typename O, typename Gen>
+    static O impl(I first, S last, O out, iter_difference_t<I> n, Gen&& g)
```

Comment:
I think here (and the other two impl functions) you can take `Gen& g`, i.e. an lvalue ref rather than a forwarding ref, and do not forward it in `operator()`.

Reply:
Done. I was wondering about that, but I decided to go with `Gen&&` because the standard says so for the `operator()`.